includes/class-coingate-for-woocommerce-payment-gateway.php

```diff
@@ -0,0 +1,426 @@
+     * @param $value
+     * @return string
+     */
+    public function validate_api_auth_token_field($key, $value) {
+        $post_data = $this->get_post_data();
+        $mode = $post_data['woocommerce_coingate_test'];
+
+        if (!empty($value)) {
+            $client = new Client();
+            $result = $client::testConnection($value, (bool)$mode);
+
+            if ($result) {
+                return $value;
+            }
+        }
+
+        WC_Admin_Settings::add_error( esc_html__( 'API Auth Token is invalid. Your changes have not been saved.', COINGATE_TRANSLATIONS ) );
+
+        return '';
+    }
+
+    /**
+     * Payment process.
+     *
+     * @param int $order_id
+     * @return string[]
+     */
+    public function process_payment($order_id) {
+        global $woocommerce, $page, $paged;
+        $order = wc_get_order($order_id);
+
+        $client = $this->init_coingate();
+
+        $description = array();
+        foreach ($order->get_items() as $item) {
+            $description[] = $item['qty'] . ' × ' . $item['name'];
+        }
+
+        $params = [
+            'order_id'          => $order->get_id(),
+            'price_amount'      => $order->get_total(),
+            'price_currency'    => $order->get_currency(),
+            'receive_currency'  => $this->receive_currency,
+            'callback_url'      => trailingslashit(get_bloginfo('wpurl')) . '?wc-api=wc_gateway_coingate',
+            'cancel_url'        => $order->get_cancel_order_url(),
+            'success_url'       => add_query_arg('order-received', $order->get_id(), add_query_arg('key', $order->get_order_key(), $this->get_return_url($order))),
+            'title'             => get_bloginfo('name', 'raw') . ' Order #' . $order->get_id(),
+            'description'       => implode(', ', $description)
+        ];
+
+        $response = ['result' => 'fail'];
+
+        try {
+            $gateway_order = $client->order->create($params);
+            if ($gateway_order) {
+                update_post_meta($order->get_id(), static::ORDER_TOKEN_META_KEY, $gateway_order->token);
+
+                $response['result'] = 'success';
+                $response['redirect'] = $gateway_order->payment_url;
+            }
+
+        } catch (ApiErrorException $exception) {
+            error_log($exception);
+        }
+
+        return $response;
+    }
+
+    /**
+     * Payment callback.
+     *
+     * @throws Exception
+     */
+    public function payment_callback() {
+        $request = $_POST;
+
+        global $woocommerce;
+
+        $order = wc_get_order($request['order_id']);
+
+        if (!$order || !$order->get_id()) {
+            throw new Exception('Order #' . $request['order_id'] . ' does not exists');
+        }
+
+        if (!$this->is_payment_valid($order, $request['token'])) {
+            throw new Exception('CoinGate callback token does not match');
+        }
+
+        $callback_order_status = $request['status'];
```

Comment:
it is important to handle the case when order payment method was changed to something else, and coingate payments overwrites the status.

Reply:
Added payment method validation. Commit: https://github.com/coingate/woocommerce-plugin/pull/9/commits/253028d1b6ef2939a00d52f9d401d4cfc7bb2460